### PR TITLE
Order Location searches using near in closest distance to center of the box

### LINKED ISSUE #4650: Feature request: Order Location searches using `near` in closest distance to center of the box
Currently, the searches on Location using `near` are not returned in the order closest to the center. It would be awesome if that could be done - either using the _sort parameter or just default to that order (aka. Manhatten distance).

### PR DESCRIPTION
**Pull request by jamesagnew**
Fix #4650

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java [hunk 1]
```diff
@@ -722,6 +724,30 @@ public boolean haveAtLeastOnePredicate() {
...
+		String latitudePlaceholder = generatePlaceholder(theLatitudeValue);
+		ComboExpression absLatitudeMiddle = new ComboExpression(ComboExpression.Op.SUBTRACT, theCoordsBuilder.getColumnLatitude(), latitudePlaceholder);
+		absLatitude = absLatitude.addCustomParams(absLatitudeMiddle);
+
+		FunctionCall absLongitude = new FunctionCall("ABS");
+		String longitudePlaceholder = generatePlaceholder(theLongitudeValue);
+		ComboExpression absLongitudeMiddle = new ComboExpression(ComboExpression.Op.SUBTRACT, theCoordsBuilder.getColumnLongitude(), longitudePlaceholder);
+		absLongitude = absLongitude.addCustomParams(absLongitudeMiddle);
+
+		ComboExpression sum = new ComboExpression(ComboExpression.Op.ADD, absLatitude, absLongitude);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java:738]: The Manhattan geometry metric.  Love it!

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceIndexedSearchParamCoords.java [hunk 2]
```diff
@@ -43,7 +43,7 @@
 @Embeddable
 @Entity
 @Table(name = "HFJ_SPIDX_COORDS", indexes = {
-	@Index(name = "IDX_SP_COORDS_HASH", columnList = "HASH_IDENTITY,SP_LATITUDE,SP_LONGITUDE"),
```
Comment (fil512) [inline hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/entity/ResourceIndexedSearchParamCoords.java:46]: I wonder how much space we'd save by dropping all our unused columns....

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4DistanceTest.java [hunk 3]
```diff
@@ -138,13 +143,202 @@ public void testNearSearchApproximateChained() {
...
+		String url = "Location?" +
+			"near2=" +
+			CoordCalculatorTestUtil.LATITUDE_CHIN + "|" +
+			CoordCalculatorTestUtil.LONGITUDE_CHIN + "|" +
+			"300" + "|" + "km" +
+			"&near=" +
+			CoordCalculatorTestUtil.LATITUDE_CHIN + "|" +
+			CoordCalculatorTestUtil.LONGITUDE_CHIN + "|" +
+			"300" + "|" + "km";
+		// FIXME: restore
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4DistanceTest.java:255]: fixme

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/util/CoordCalculatorTestUtil.java [hunk 4]
```diff
@@ -33,6 +33,14 @@ public final class CoordCalculatorTestUtil {
 	public static final double LONGITIDE_TAVEUNI = 179.889793;
 	// enough distance from point to cross anti-meridian
 	public static final double DISTANCE_TAVEUNI = 100.0;
+	public static final double LATITUDE_TORONTO = 43.741667;
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/util/CoordCalculatorTestUtil.java:36]: haha.  this kind of work is so much fun...

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 5]
```diff
@@ -865,6 +866,12 @@ private void createSort(QueryStack theQueryStack, SortSpec theSort) {
 
 					break;
 				case SPECIAL:
+					if (param.getPath().equals(LOCATION_POSITION)) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:869]: fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/predicate/CoordsPredicateBuilder.java [hunk 6]
```diff
@@ -112,10 +78,7 @@ public Condition createPredicateCoords(SearchParameterMap theParams,
 		} else if (distanceKm > CoordCalculator.MAX_SUPPORTED_DISTANCE_KM) {
 			throw new IllegalArgumentException(Msg.code(1234) + "Invalid " + Location.SP_NEAR_DISTANCE + " parameter '" + distanceKm + "' must be <= " + CoordCalculator.MAX_SUPPORTED_DISTANCE_KM);
 		} else {
-			double latitudeDegrees = Double.parseDouble(latitudeValue);
-			double longitudeDegrees = Double.parseDouble(longitudeValue);
-
-			GeoBoundingBox box = CoordCalculator.getBox(latitudeDegrees, longitudeDegrees, distanceKm);
+			GeoBoundingBox box = CoordCalculator.getBox((latitudeValue), (longitudeValue), distanceKm);
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/predicate/CoordsPredicateBuilder.java:81]: fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java [hunk 7]
```diff
@@ -722,6 +724,30 @@ public boolean haveAtLeastOnePredicate() {
 		return myHaveAtLeastOnePredicate;
 	}
 
+	public void addSortCoordsNear(CoordsPredicateBuilder theCoordsBuilder, double theLatitudeValue, double theLongitudeValue, boolean theAscending) {
+		FunctionCall absLatitude = new FunctionCall("ABS");
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/sql/SearchQueryBuilder.java:728]: It's not a part of the official sql92 spec, but it looks like all of our supported DBs support it.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/util/CoordCalculatorTestUtil.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/util/CoordCalculatorTestUtil.java:36]: right??

**Pull request merged**